README.rst
============
txWS Upgrade
============

txWS-Upgrade (pronounced "Twisted WebSockets Upgrade") is a small, short, simple library for
adding WebSockets server support to your favorite Twisted applications.

This is forked from txWS to add upgrade support.
https://github.com/MostAwesomeDude/txWS


Usage
=====

Use ``txws.WebSocketFactory`` to wrap your factories. That's it! Adding
WebSockets support has never been easier.

    >>> from txws import WebSocketFactory
    >>> reactor.listenTCP(8080, WebSocketFactory(factory_to_wrap))

There is no extra trick to txWS. There is no special setup involved.

Do you want secure WebSockets? Use ``listenSSL()`` instead of ``listenTCP()``.

Upgrade Usage
=============

If you want to use the websocket with in an existing site, Update your code as follows.

This is a vanilla Twisted website. ::


        from twisted.web import server
        from twisted.web.resource import Resource
        from twisted.internet import reactor, endpoints

        class Simple(Resource):
            isLeaf = True
            def getChild(self, name, request):
                if name == '':
                    return self
                return Resource.getChild(self, name, request)

            def render_GET(self, request):
                return "Hello, world! I am located at %r." % (request.prepath,)

        rootResource = Simple()
        site = server.Site(rootResource)

        endpoint = endpoints.TCP4ServerEndpoint(reactor, 8080)
        endpoint.listen(site)
        reactor.run()



Now add the website has support for WebSockets, to include
 the ``txws.WebSocketUpgradeResource`` and ``txws.WebSocketUpgradeHTTPChannel``. ::

        from twisted.web import server
        from twisted.web.resource import Resource
        from twisted.internet import reactor, endpoints

        class Simple(Resource):
            isLeaf = True
            def getChild(self, name, request):
                if name == '':
                    return self
                return Resource.getChild(self, name, request)

            def render_GET(self, request):
                return "Hello, world! I am located at %r." % (request.prepath,)

        rootResource = Simple()
        site = server.Site(rootResource)

        # 1) Add the imports
        #    Create the WebSocketFactory
        #    Create the WebSocketUpgradeResource
        #    Put the resource into the resource tree
        from txws import WebSocketFactory, WebSocketUpgradeResource
        rootResource.putChild(b"websocket",
                              WebSocketUpgradeResource(WebSocketFactory(factory_to_wrap)))


        # 2) Add the imports
        #    Replace protocol for the website with the Websocket upgradable ones
        from txws import WebSocketUpgradeHTTPChannel
        site.protocol = VortexWebsocketHTTPChannel

        endpoint = endpoints.TCP4ServerEndpoint(reactor, 8080)
        endpoint.listen(site)
        reactor.run()


Versions
========

txWS supports the following versions of the WebSockets draft:

 * Version 76

   * Hixie-76 (Chrome 6, Fx 4, Opera 11, **UNTESTED** Safari 5)
   * HyBi-00

 * Version 7

   * HyBi-07 (Fx 6)

 * Version 8

   * HyBi-08
   * HyBi-10 (Chrome 14, Chrome 15, Fx 7, Fx 8)

 * Version 13

   * RFC 6455 (Chrome 16)

All listed browser versions have been tested and verified working; any browser
marked "UNTESTED" hasn't been personally tested, but has been reported working
by third parties.

In case you're wondering, the version numbers above are correct; WebSockets
versioning is not sane.

Browser Quirks
==============

This might save you some time when developing your WebSockets-based
application.

 * Firefox (all versions): WebSockets do not follow the standard WebSocket
   API.
 * Opera 11: WebSockets are disabled by default and are very slow to close
   connections.

Comparisons
===========

Here's how txWS compares to other Twisted WebSockets libraries.

txWebSockets
------------

txWS, unlike txWebSockets, doesn't reuse any HTTP machinery and doesn't
pretend to be HTTP. Whether this is a good or bad thing depends largely on
whether the WebSockets standard ends up being a valid HTTP subset.

txWS supports newer WS versions 7 and 8, but txWebSockets supports the older
version 75. Both libraries support version 76.

Autobahn
--------

Autobahn provides a client library for WebSockets as well as a server, and
provides a fancy set of messaging protocols on top of the WS layer. Autobahn
also provides support for WS version 10.

However, Autobahn doesn't provide support for WS version 76, and requires
clients to subclass their factories and protocols in order to provide WS
functionality. txWS uses a compositional approach with wrapped protocols,
allowing completely transparent reuse of existing protocols and factories.

Cyclone
-------

Cyclone provides a simple WebSockets handler. This handler can do WS versions
75 and 76. The Cyclone WebSockets handler is very limited, can only wrap other
Cyclone handlers, and doesn't support any of the more modern WebSockets
versions.

License
=======

txWS is (c) 2011 Oregon State University Open Source Lab, (c) 2014 Google
Inc., and is made available under the Apache 2.0 license.

Thanks
======

Thank you to all of the contributors in the community who have chipped in to
help keep txWS alive.


CHANGELOG.rst
0.10
===

Support for HTTP upgrades within Twisted

0.9
===

* HyBi-07 updates for Chrome and Fx (#18)
* Python 3 forward-compatibility improvements (#23)

0.8
===

* Add binary message support (#15, #19)
* Negotiate multiple subprotocols (#21)

0.7
===

* Fix ITransport.writeSequence() implementation (#3)
* Add RFC 6455 (Chrome 16) support (#4)

0.6.2
=====

* Add documentation on differences between txWS, txWebSockets, and Autobahn
  (#1)
* Add HyBi-10 compatibility (Chrome 14, Firefox 7)
* Support for noVNC

  * Fix partial frame decoding for HyBi-00
  * Fix partial frame decoding for HyBi-07
  * Fix jumbo frame lengths for HyBi-07
  * Fix early frames not being sent in all cases

0.6.1
=====

* Fix issue with installation from PyPI

0.6
===

* Start of availability on PyPI

0.5
===

* Initial release


tests.py
# Copyright 2014 Google Inc. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
import six
from twisted.trial import unittest

from txws import (is_hybi00, complete_hybi00, make_hybi00_frame,
                  parse_hybi00_frames, http_headers, make_accept, mask, CLOSE,
                  NORMAL, PING, PONG, parse_hybi07_frames)

class TestHTTPHeaders(unittest.TestCase):

    def test_single_header(self):
        raw = "Connection: Upgrade"
        headers = http_headers(raw)
        self.assertTrue("Connection" in headers)
        self.assertEqual(headers["Connection"], "Upgrade")

    def test_single_header_newline(self):
        raw = "Connection: Upgrade\r\n"
        headers = http_headers(raw)
        self.assertEqual(headers["Connection"], "Upgrade")

    def test_multiple_headers(self):
        raw = "Connection: Upgrade\r\nUpgrade: WebSocket"
        headers = http_headers(raw)
        self.assertEqual(headers["Connection"], "Upgrade")
        self.assertEqual(headers["Upgrade"], "WebSocket")

    def test_origin_colon(self):
        """
        Some headers have multiple colons in them.
        """

        raw = "Origin: http://example.com:8080"
        headers = http_headers(raw)
        self.assertEqual(headers["Origin"], "http://example.com:8080")

class TestKeys(unittest.TestCase):

    def test_make_accept_rfc(self):
        """
        Test ``make_accept()`` using the keys listed in the RFC for HyBi-07
        through HyBi-10.
        """

        key = "dGhlIHNhbXBsZSBub25jZQ=="

        self.assertEqual(make_accept(key), "s3pPLMBiTxaQ9kYGzzhZRbK+xOo=")

    def test_make_accept_wikipedia(self):
        """
        Test ``make_accept()`` using the keys listed on Wikipedia.
        """

        key = "x3JJHMbDL1EzLkh9GBhXDw=="

        self.assertEqual(make_accept(key), "HSmrc0sMlYUkAGmm5OPpG2HaGWk=")

class TestHyBi00(unittest.TestCase):

    def test_is_hybi00(self):
        headers = {
            "Sec-WebSocket-Key1": "hurp",
            "Sec-WebSocket-Key2": "derp",
        }
        self.assertTrue(is_hybi00(headers))

    def test_is_hybi00_false(self):
        headers = {
            "Sec-WebSocket-Key1": "hurp",
        }
        self.assertFalse(is_hybi00(headers))

    def test_complete_hybi00_wikipedia(self):
        """
        Test complete_hybi00() using the keys listed on Wikipedia's WebSockets
        page.
        """

        headers = {
            "Sec-WebSocket-Key1": "4 @1  46546xW%0l 1 5",
            "Sec-WebSocket-Key2": "12998 5 Y3 1  .P00",
        }
        challenge = "^n:ds[4U"

        self.assertEqual(complete_hybi00(headers, challenge),
                         six.b("8jKS'y:G*Co,Wxa-"))

    def test_make_hybi00(self):
        """
        HyBi-00 frames are really, *really* simple.
        """

        self.assertEqual(make_hybi00_frame("Test!"), six.b("\x00Test!\xff"))

    def test_parse_hybi00_single(self):
        frame = six.b("\x00Test\xff")

        frames, buf = parse_hybi00_frames(frame)

        self.assertEqual(len(frames), 1)
        self.assertEqual(frames[0], (NORMAL, six.b("Test")))
        self.assertEqual(buf, six.b(""))

    def test_parse_hybi00_multiple(self):
        frame = six.b("\x00Test\xff\x00Again\xff")

        frames, buf = parse_hybi00_frames(frame)

        self.assertEqual(len(frames), 2)
        self.assertEqual(frames[0], (NORMAL, six.b("Test")))
        self.assertEqual(frames[1], (NORMAL, six.b("Again")))
        self.assertEqual(buf, six.b(""))

    def test_parse_hybi00_incomplete(self):
        frame = six.b("\x00Test")

        frames, buf = parse_hybi00_frames(frame)

        self.assertEqual(len(frames), 0)
        self.assertEqual(buf, six.b("\x00Test"))

    def test_parse_hybi00_garbage(self):
        frame = six.b("trash\x00Test\xff")

        frames, buf = parse_hybi00_frames(frame)

        self.assertEqual(len(frames), 1)
        self.assertEqual(frames[0], (NORMAL, six.b("Test")))
        self.assertEqual(buf, six.b(""))

    def test_socketio_crashers(self):
        """
        A series of snippets which crash other WebSockets implementations
        (specifically, Socket.IO) are harmless to this implementation.
        """

        frames = [
            """[{"length":1}]""",
            """{"messages":[{"length":1}]}""",
            "hello",
            "hello<script>alert(/xss/)</script>",
            "hello<img src=x:x onerror=alert(/xss.2/)>",
            "{",
            "~m~EVJLFDJP~",
        ]

        for frame in frames:
            prepared = make_hybi00_frame(frame)
            frames, buf = parse_hybi00_frames(prepared)

            self.assertEqual(len(frames), 1)
            self.assertEqual(frames[0], (NORMAL, frame.encode('utf-8')))
            self.assertEqual(buf, six.b(""))

class TestHyBi07Helpers(unittest.TestCase):
    """
    HyBi-07 is best understood as a large family of helper functions which
    work together, somewhat dysfunctionally, to produce a mediocre
    Thanksgiving every other year.
    """

    def test_mask_noop(self):
        key = six.b("\x00\x00\x00\x00")
        self.assertEqual(mask(six.b("Test"), key), six.b("Test"))

    def test_mask_noop_long(self):
        key = six.b("\x00\x00\x00\x00")
        self.assertEqual(mask(six.b("LongTest"), key), six.b("LongTest"))

    def test_parse_hybi07_unmasked_text(self):
        """
        From HyBi-10, 4.7.
        """

        frame = six.b("\x81\x05Hello")
        frames, buf = parse_hybi07_frames(frame)
        self.assertEqual(len(frames), 1)
        self.assertEqual(frames[0], (NORMAL, six.b("Hello")))
        self.assertEqual(buf, six.b(""))

    def test_parse_hybi07_masked_text(self):
        """
        From HyBi-10, 4.7.
        """

        frame = six.b("\x81\x857\xfa!=\x7f\x9fMQX")
        frames, buf = parse_hybi07_frames(frame)
        self.assertEqual(len(frames), 1)
        self.assertEqual(frames[0], (NORMAL, six.b("Hello")))
        self.assertEqual(buf, six.b(""))

    def test_parse_hybi07_unmasked_text_fragments(self):
        """
        We don't care about fragments. We are totally unfazed.

        From HyBi-10, 4.7.
        """

        frame = six.b("\x01\x03Hel\x80\x02lo")
        frames, buf = parse_hybi07_frames(frame)
        self.assertEqual(len(frames), 2)
        self.assertEqual(frames[0], (NORMAL, six.b("Hel")))
        self.assertEqual(frames[1], (NORMAL, six.b("lo")))
        self.assertEqual(buf, six.b(""))

    def test_parse_hybi07_ping(self):
        """
        From HyBi-10, 4.7.
        """

        frame = six.b("\x89\x05Hello")
        frames, buf = parse_hybi07_frames(frame)
        self.assertEqual(len(frames), 1)
        self.assertEqual(frames[0], (PING, six.b("Hello")))
        self.assertEqual(buf, six.b(""))

    def test_parse_hybi07_pong(self):
        """
        From HyBi-10, 4.7.
        """

        frame = six.b("\x8a\x05Hello")
        frames, buf = parse_hybi07_frames(frame)
        self.assertEqual(len(frames), 1)
        self.assertEqual(frames[0], (PONG, six.b("Hello")))
        self.assertEqual(buf, six.b(""))

    def test_parse_hybi07_close_empty(self):
        """
        A HyBi-07 close packet may have no body. In that case, it should use
        the generic error code 1000, and have no reason.
        """

        frame = six.b("\x88\x00")
        frames, buf = parse_hybi07_frames(frame)
        self.assertEqual(len(frames), 1)
        self.assertEqual(frames[0], (CLOSE, (1000, six.b("No reason given"))))
        self.assertEqual(buf, six.b(""))

    def test_parse_hybi07_close_reason(self):
        """
        A HyBi-07 close packet must have its first two bytes be a numeric
        error code, and may optionally include trailing text explaining why
        the connection was closed.
        """

        frame = six.b("\x88\x0b\x03\xe8No reason")
        frames, buf = parse_hybi07_frames(frame)
        self.assertEqual(len(frames), 1)
        self.assertEqual(frames[0], (CLOSE, (1000, six.b("No reason"))))
        self.assertEqual(buf, six.b(""))

    def test_parse_hybi07_partial_no_length(self):
        frame = six.b("\x81")
        frames, buf = parse_hybi07_frames(frame)
        self.assertFalse(frames)
        self.assertEqual(buf, six.b("\x81"))

    def test_parse_hybi07_partial_truncated_length_int(self):
        frame = six.b("\x81\xfe")
        frames, buf = parse_hybi07_frames(frame)
        self.assertFalse(frames)
        self.assertEqual(buf, six.b("\x81\xfe"))

    def test_parse_hybi07_partial_truncated_length_double(self):
        frame = six.b("\x81\xff")
        frames, buf = parse_hybi07_frames(frame)
        self.assertFalse(frames)
        self.assertEqual(buf, six.b("\x81\xff"))

    def test_parse_hybi07_partial_no_data(self):
        frame = six.b("\x81\x05")
        frames, buf = parse_hybi07_frames(frame)
        self.assertFalse(frames)
        self.assertEqual(buf, six.b("\x81\x05"))

    def test_parse_hybi07_partial_truncated_data(self):
        frame = six.b("\x81\x05Hel")
        frames, buf = parse_hybi07_frames(frame)
        self.assertFalse(frames)
        self.assertEqual(buf, six.b("\x81\x05Hel"))


tests/test_websocket_new.py
from twisted.trial import unittest

from twisted.protocols.policies import ProtocolWrapper
from vncap.websocket_new import (WebSocketProtocol, complete_hybi00,
                                 http_headers, FRAMES)

class TestHTTPHeaders(unittest.TestCase):

    def test_single_header(self):
        raw = "Connection: Upgrade"
        headers = http_headers(raw)
        self.assertTrue("Connection" in headers)
        self.assertEqual(headers["Connection"], "Upgrade")

    def test_single_header_newline(self):
        raw = "Connection: Upgrade\r\n"
        headers = http_headers(raw)
        self.assertEqual(headers["Connection"], "Upgrade")

    def test_multiple_headers(self):
        raw = "Connection: Upgrade\r\nUpgrade: WebSocket"
        headers = http_headers(raw)
        self.assertEqual(headers["Connection"], "Upgrade")
        self.assertEqual(headers["Upgrade"], "WebSocket")

    def test_origin_colon(self):
        """
        Some headers have multiple colons in them.
        """

        raw = "Origin: http://example.com:8080"
        headers = http_headers(raw)
        self.assertEqual(headers["Origin"], "http://example.com:8080")

class TestHyBi00(unittest.TestCase):

    def test_complete_hybi00_wikipedia(self):
        """
        Test complete_hybi00() using the keys listed on Wikipedia's WebSockets
        page.
        """

        headers = {
            "Sec-WebSocket-Key1": "4 @1  46546xW%0l 1 5",
            "Sec-WebSocket-Key2": "12998 5 Y3 1  .P00",
        }
        challenge = "^n:ds[4U"

        self.assertEqual(complete_hybi00(headers, challenge),
                         "8jKS'y:G*Co,Wxa-")

class TestWebSocketProtocolFrames(unittest.TestCase):

    def setUp(self):
        self.proto = WebSocketProtocol(None, None)
        self.proto.state = FRAMES

        self.expected = []
        def dr(chaff, data):
            self.expected.append(data)
        self.patch(ProtocolWrapper, "dataReceived", dr)

        self.sent = []
        class FakeTransport:
            def write(chaff, data):
                self.sent.append(data)
        self.proto.transport = FakeTransport()

    def test_trivial(self):
        pass

    def test_parse_frames_single(self):
        frame = "\x00Test\xff"

        self.proto.buf = frame
        self.proto.parse_frames()

        self.assertEqual(len(self.expected), 1)
        self.assertEqual(self.expected[0], "Test")

    def test_parse_frames_multiple(self):
        frame = "\x00Test\xff\x00Again\xff"

        self.proto.buf = frame
        self.proto.parse_frames()

        self.assertEqual(len(self.expected), 2)
        self.assertEqual(self.expected[0], "Test")
        self.assertEqual(self.expected[1], "Again")

    def test_parse_frames_incomplete(self):
        frame = "\x00Test"

        self.proto.buf = frame
        self.proto.parse_frames()

        self.assertEqual(len(self.expected), 0)

    def test_parse_frames_garbage(self):
        frame = "trash\x00Test\xff"

        self.proto.buf = frame
        self.proto.parse_frames()

        self.assertEqual(len(self.expected), 1)
        self.assertEqual(self.expected[0], "Test")

    def test_send_frames_multiple(self):
        self.proto.pending_frames.append("hello")
        self.proto.pending_frames.append("world")

        self.proto.send_frames()
        self.assertEqual(len(self.sent), 2)
        self.assertEqual(self.sent[0], "\x00hello\xff")
        self.assertEqual(self.sent[1], "\x00world\xff")


site.py
from base64 import b64encode, b64decode
from functools import partial

from twisted.web.resource import Resource
from twisted.protocols.policies import ProtocolWrapper
from twisted.python import log

from vncap.protocol import make_server_and_client
from vncap.websocket import WebSocketSite, WebSocketHandler

class DummyFactory(object):
    """
    A shim to fool ProtocolWrapper.

    We don't actually care about the wrapper factory functionality, so.
    """

    def registerProtocol(self, protocol):
        pass

    def unregisterProtocol(self, protocol):
        pass

class Base64Transport(ProtocolWrapper):

    def dataReceived(self, data):
        ProtocolWrapper.dataReceived(self, b64decode(data))

    def write(self, data):
        ProtocolWrapper.write(self, b64encode(data))

    def writeSequence(self, data):
        ProtocolWrapper.writeSequence(self, [b64encode(data) for i in data])

class VNCHandler(WebSocketHandler):
    """
    A handler that pretends the other side of the connection is VNC over WS.

    Specifically, the other side is probably NoVNC, which would like us to
    base64-encode our data.
    """

    def __init__(self, transport, host="", port=0, password=""):
        WebSocketHandler.__init__(self, transport)
        server = make_server_and_client(host, port, password)
        self.wrapped = Base64Transport(DummyFactory(), server)
        self.transport = transport

    def connectionMade(self):
        self.wrapped.makeConnection(self.transport)

    def frameReceived(self, data):
        self.wrapped.dataReceived(data)

class VNCSite(WebSocketSite):

    def __init__(self, host, port, password):
        handler = partial(VNCHandler, host=host, port=port, password=password)
        resource = Resource()
        WebSocketSite.__init__(self, resource)
        self.addHandler("/", handler)


websocket.py
# Copyright (c) 2011 Oregon State University

"""
Blind reimplementation of WebSockets as a standalone wrapper for Twisted
protocols.
"""

from base64 import b64encode, b64decode
from hashlib import md5, sha1
from string import digits
from struct import pack, unpack

from twisted.internet.interfaces import ISSLTransport
from twisted.protocols.policies import ProtocolWrapper, WrappingFactory
from twisted.python import log
from twisted.web.http import datetimeToString

# Flavors of WS supported here.
# HYBI00 - Hixie-76, HyBi-00. Challenge/response after headers, very minimal
#          framing. Tricky to start up, but very smooth sailing afterwards.
# HYBI06 - HyBi-06, HyBi-07. Modern "standard" handshake. Bizarre masked
#          frames, lots of binary data packing.

HYBI00, HYBI06 = range(2)

# States of the state machine. Because there are no reliable byte counts for
# any of this, we don't use StatefulProtocol; instead, we use custom state
# enumerations. Yay!

REQUEST, NEGOTIATING, CHALLENGE, FRAMES = range(4)

# Control frame specifiers. Some versions of WS have control signals sent
# in-band. Adorable, right?

NORMAL, CLOSE, PING, PONG = range(4)

opcode_types = {
    0x0: NORMAL,
    0x1: NORMAL,
    0x2: NORMAL,
    0x8: CLOSE,
    0x9: PING,
    0xa: PONG,
}

encoders = {
    "base64": b64encode,
}

decoders = {
    "base64": b64decode,
}

# Fake HTTP stuff, and a couple convenience methods for examining fake HTTP
# headers.

def http_headers(s):
    """
    Create a dictionary of data from raw HTTP headers.
    """

    d = {}

    for line in s.split("\r\n"):
        try:
            key, value = [i.strip() for i in line.split(":", 1)]
            d[key] = value
        except ValueError:
            pass

    return d

def is_websocket(headers):
    """
    Determine whether a given set of headers is asking for WebSockets.
    """

    return ("Upgrade" in headers.get("Connection", "")
            and headers.get("Upgrade").lower() == "websocket")

def is_hybi00(headers):
    """
    Determine whether a given set of headers is HyBi-00-compliant.

    Hixie-76 and HyBi-00 use a pair of keys in the headers to handshake with
    servers.
    """

    return "Sec-WebSocket-Key1" in headers and "Sec-WebSocket-Key2" in headers

# Authentication for WS.

def complete_hybi00(headers, challenge):
    """
    Generate the response for a HyBi-00 challenge.
    """

    key1 = headers["Sec-WebSocket-Key1"]
    key2 = headers["Sec-WebSocket-Key2"]

    first = int("".join(i for i in key1 if i in digits)) / key1.count(" ")
    second = int("".join(i for i in key2 if i in digits)) / key2.count(" ")

    nonce = pack(">II8s", first, second, challenge)

    return md5(nonce).digest()

def make_accept(key):
    """
    Create an "accept" response for a given key.

    This dance is expected to somehow magically make WebSockets secure.
    """

    guid = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

    return sha1("%s%s" % (key, guid)).digest().encode("base64").strip()

# Frame helpers.
# Separated out to make unit testing a lot easier.
# Frames are bonghits in newer WS versions, so helpers are appreciated.

def make_hybi00_frame(buf):
    """
    Make a HyBi-00 frame from some data.

    This function does exactly zero checks to make sure that the data is safe
    and valid text without any 0xff bytes.
    """

    return "\x00%s\xff" % buf

def parse_hybi00_frames(buf):
    """
    Parse HyBi-00 frames, returning unwrapped frames and any unmatched data.

    This function does not care about garbage data on the wire between frames,
    and will actively ignore it.
    """

    start = buf.find("\x00")
    frames = []

    while start != -1:
        end = buf.find("\xff")
        if end == -1:
            # Incomplete frame, try again later.
            return frames, buf
        else:
            frame, buf = buf[start + 1:end], buf[end + 1:]
            # Found a frame, put it in the list.
            frames.append((NORMAL, frame))
        start = buf.find("\x00")

    # We appear to have hit an exact frame boundary. This is actually pretty
    # likely with a lot of implementations.
    return frames, buf

def mask(buf, key):
    """
    Mask or unmask a buffer of bytes with a masking key.

    The key must be exactly four bytes long.
    """

    # This is super-secure, I promise~
    key = [ord(i) for i in key]
    buf = list(buf)
    for i, char in enumerate(buf):
        buf[i] = chr(ord(char) ^ key[i % 4])
    return "".join(buf)

def make_hybi06_frame(buf):
    """
    Make a HyBi-06 frame.
    """

    # XXX next commit
    return None

def parse_hybi06_frames(buf):
    """
    Parse HyBi-06 frames in a highly compliant manner.

    This function requires *complete* frames and does highly bogus things if
    fed incomplete frames.
    """

    frames = []

    while buf:
        # Grab the header. This single byte holds some flags nobody cares
        # about, and an opcode which nobody cares about.
        header, buf = ord(buf[0]), buf[1:]
        if header & 0x70:
            # At least one of the reserved flags is set. Pork chop sandwiches!
            raise Exception("Reserved flag in HyBi-06 frame (%d)" % header)
            frames.append(("", CLOSE))
            return frames, buf

        # Get the opcode, and translate it to a local enum which we actually
        # care about.
        opcode = header & 0xf
        try:
            opcode = opcode_types[opcode]
        except KeyError:
            raise Exception("Unknown opcode %d in HyBi-06 frame!" % opcode)

        # Get the payload length and determine whether we need to look for an
        # extra length.
        length, buf = ord(buf[0]), buf[1:]
        masked = length & 0x80
        length &= 0x7f

        # Extra length fields.
        if length == 0x7e:
            length, buf = buf[:4], buf[4:]
            length = unpack(">H", length)[0]
        elif length == 0x7f:
            # Protocol bug: The top bit of this long long *must* be cleared;
            # that is, it is expected to be interpreted as signed. That's
            # fucking stupid, if you don't mind me saying so, and so we're
            # interpreting it as unsigned anyway. If you wanna send exabytes
            # of data down the wire, then go ahead!
            length, buf = buf[:8], buf[8:]
            length = unpack(">Q", length)[0]

        if masked:
            key, buf = buf[:4], buf[4:]

        data, buf = buf[:length], buf[length:]

        if masked:
            data = mask(data, key)

        if opcode == CLOSE:
            # Gotta unpack the opcode and return usable data here.
            data = unpack(">H", data[:2])[0], data[2:]

        frames.append((opcode, data))

    return frames, buf

class WebSocketProtocol(ProtocolWrapper):
    """
    Protocol which wraps another protocol to provide a WebSockets transport
    layer.
    """

    buf = ""
    codec = None
    location = "/"
    host = "example.com"
    origin = "http://example.com"
    state = REQUEST
    flavor = None

    def __init__(self, *args, **kwargs):
        ProtocolWrapper.__init__(self, *args, **kwargs)
        self.pending_frames = []

    def isSecure(self):
        """
        Borrowed technique for determining whether this connection is over
        SSL/TLS.
        """

        return ISSLTransport(self.transport, None) is not None

    def sendCommonPreamble(self):
        """
        Send the preamble common to all WebSockets connections.

        This might go away in the future if WebSockets continue to diverge.
        """

        self.transport.writeSequence([
            "HTTP/1.1 101 FYI I am not a webserver\r\n",
            "Server: TwistedWebSocketWrapper/1.0\r\n",
            "Date: %s\r\n" % datetimeToString(),
            "Upgrade: WebSocket\r\n",
            "Connection: Upgrade\r\n",
        ])

    def sendHyBi00Preamble(self):
        """
        Send a HyBi-00 preamble.
        """

        protocol = "wss" if self.isSecure() else "ws"

        self.sendCommonPreamble()

        self.transport.writeSequence([
            "Sec-WebSocket-Origin: %s\r\n" % self.origin,
            "Sec-WebSocket-Location: %s://%s%s\r\n" % (protocol, self.host,
                                                       self.location),
            "WebSocket-Protocol: %s\r\n" % self.codec,
            "Sec-WebSocket-Protocol: %s\r\n" % self.codec,
            "\r\n",
        ])

    def sendHyBi06Preamble(self):
        """
        Send a HyBi-06 preamble.
        """

        self.sendCommonPreamble()

        self.transport.write("Sec-WebSocket-Accept: %s\r\n\r\n" %
                             make_accept(self.headers))

    def parseFrames(self):
        """
        Find frames in incoming data and pass them to the underlying protocol.
        """

        if self.flavor == HYBI00:
            parser = parse_hybi00_frames
        elif self.flavor == HYBI06:
            parser = parse_hybi06_frames
        else:
            raise Exception("Unknown flavor %r!" % self.flavor)

        frames, self.buf = parser(self.buf)

        for frame in frames:
            opcode = frame[0]
            if opcode == NORMAL:
                # Business as usual. Decode the frame, if we have a decoder.
                if self.codec:
                    frame = decoders[self.codec](frame)
                # Pass the frame to the underlying protocol.
                ProtocolWrapper.dataReceived(self, frame)
            elif opcode == CLOSE:
                # The other side wants us to close. I wonder why?
                reason, text = frame[1]
                log.msg("Closing connection: %r (%d)" % (text, reason))

                # Send the smallest possible closing message.
                # XXX totally protocol-7 specific
                self.transport.write("\x88\x00")

    def sendFrames(self):
        """
        Send all pending frames.
        """

        if self.state != FRAMES:
            return

        if self.flavor == HYBI00:
            maker = make_hybi00_frame
        elif self.flavor == HYBI06:
            maker = make_hybi06_frame
        else:
            raise Exception("Unknown flavor %r!" % self.flavor)

        for frame in self.pending_frames:
            # Encode the frame before sending it.
            if self.codec:
                frame = encoders[self.codec](frame)
            packet = maker(frame)
            self.transport.write(packet)
        self.pending_frames = []

    def validateHeaders(self):
        """
        Check received headers for sanity and correctness, and stash any data
        from them which will be required later.
        """

        # Obvious but necessary.
        if not is_websocket(self.headers):
            log.msg("Not handling non-WS request")
            return False

        # Stash host and origin for those browsers that care about it.
        if "Host" in self.headers:
            self.host = self.headers["Host"]
        if "Origin" in self.headers:
            self.origin = self.headers["Origin"]

        # Check whether a codec is needed. WS calls this a "protocol" for
        # reasons I cannot fathom.
        protocol = None
        if "WebSocket-Protocol" in self.headers:
            protocol = self.headers["WebSocket-Protocol"]
        elif "Sec-WebSocket-Protocol" in self.headers:
            protocol = self.headers["Sec-WebSocket-Protocol"]

        if protocol:
            if protocol not in encoders or protocol not in decoders:
                log.msg("Couldn't handle WS protocol %s!" % protocol)
                return False
            self.codec = protocol

        # Start the next phase of the handshake for HyBi-00.
        if is_hybi00(self.headers):
            log.msg("Starting HyBi-00/Hixie-76 handshake")
            self.flavor = HYBI00
            self.state = CHALLENGE

        # Start the next phase of the handshake for HyBi-06+.
        if "Sec-WebSocket-Version" in self.headers:
            log.msg("Starting HyBi-06+ (v6, v7) conversation")
            self.sendHyBi06Preamble()
            self.flavor = HYBI06
            self.state = FRAMES

        return True

    def dataReceived(self, data):
        self.buf += data

        log.msg("Got data, buffer is %r" % self.buf)

        oldstate = None

        while oldstate != self.state:
            oldstate = self.state

            # Handle initial requests. These look very much like HTTP
            # requests, but aren't. We need to capture the request path for
            # those browsers which want us to echo it back to them (Chrome,
            # mainly.)
            # These lines look like:
            # GET /some/path/to/a/websocket/resource HTTP/1.1
            if self.state == REQUEST:
                if "\r\n" in self.buf:
                    request, chaff, self.buf = self.buf.partition("\r\n")
                    try:
                        verb, self.location, version = request.split(" ")
                    except ValueError:
                        self.loseConnection()
                    else:
                        self.state = NEGOTIATING

            elif self.state == NEGOTIATING:
                # Check to see if we've got a complete set of headers yet.
                if "\r\n\r\n" in self.buf:
                    head, chaff, self.buf = self.buf.partition("\r\n\r\n")
                    self.headers = http_headers(head)
                    # Validate headers. This will cause a state change.
                    if not self.validateHeaders():
                        self.loseConnection()

            elif self.state == CHALLENGE:
                # Handle the challenge. This is completely exclusive to
                # HyBi-00/Hixie-76.
                if len(self.buf) >= 8:
                    challenge, self.buf = self.buf[:8], self.buf[8:]
                    response = complete_hybi00(self.headers, challenge)
                    self.sendHyBi00Preamble()
                    self.transport.write(response)
                    log.msg("Completed HyBi-00/Hixie-76 handshake")
                    # Start sending frames, and kick any pending frames.
                    self.state = FRAMES
                    self.sendFrames()

            elif self.state == FRAMES:
                self.parseFrames()

    def write(self, data):
        """
        Write to the transport.

        This method will only be called by the underlying protocol.
        """

        self.pending_frames.append(data)
        self.sendFrames()

class WebSocketFactory(WrappingFactory):
    """
    Factory which wraps another factory to provide WebSockets transports for
    all of its protocols.
    """

    protocol = WebSocketProtocol
